### PR TITLE
Plugin Docs

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds docs and api references for our new plugin interface.
Adds a diagram (svg image) with an overview of how accelerators and plugins are composed.

Note: Had to convert some typing in the signatures because sphinx auto doc does not support cyclic forward references.

overview

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/sharded.py [hunk 1]
```diff
@@ -30,6 +30,7 @@
 
 
 class DDPShardedPlugin(DDPPlugin):
+    """ Optimizer sharded training provided by FairScale. """
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/sharded.py:33]:
```suggestion
    """ Optimizer and gradient sharded training provided by FairScale. """
```

**Review by ananthsub — APPROVED** [review]

File: docs/source/extensions/plugins.rst [hunk 2]
```diff
@@ -4,6 +4,135 @@
...
+.. code-block:: python
+
+    # accelerator: GPUAccelerator
+    # training type: DDPPlugin
+    # precision: NativeMixedPrecisionPlugin
+    trainer = Trainer(gpus=4, precision=16)
+
+There are two types of Plugins in Lightning with different responsibilities:
+
+TrainingTypePlugin:
```
Comment (edenlightning) [inline docs/source/extensions/plugins.rst:23]:
```suggestion
TrainingTypePlugin
==============
```

File: docs/source/extensions/plugins.rst [hunk 3]
```diff
@@ -4,6 +4,135 @@
...
+
+- Launching and teardown of training processes (if applicable)
+- Setup communication between processes (NCCL, GLOO, MPI, ...)
+- Provide a unified communication interface for reduction, broadcast, etc.
+- Provide access to the wrapped LightningModule
+
+PrecisionPlugin:
```
Comment (edenlightning) [inline docs/source/extensions/plugins.rst:30]:
```suggestion
PrecisionPlugin
===========
```

File: docs/source/extensions/plugins.rst [hunk 4]
```diff
@@ -4,6 +4,135 @@
 Plugins
 #######
 
-Plugins allow custom integrations to the internals of the Trainer such as a custom amp or ddp implementation.
+Plugins allow custom integrations to the internals of the Trainer such as a custom precision or
+distributed implementation.
+
+We expose Accelerators and Plugins mainly for expert users who want to extend Lightning to work with new
```
Comment (edenlightning) [inline docs/source/extensions/plugins.rst:10]: Maybe

"Under the hood, the Lightning Trainer is using plugins in the training routine, added automatically depending on the provided Trainer arguments. For example:
.....

We expose Accelerators and Plugins mainly for expert users that want to extend Lightning for:
* New hardware (like TPU plugin)
* Distributed training algorithms (such as ...)
* Clusters (such as ....)

File: docs/source/extensions/plugins.rst [hunk 5]
```diff
@@ -4,6 +4,135 @@
...
+PrecisionPlugin:
+
+- Perform pre- and post backward/optimizer step operations such as scaling gradients
+- Provide context managers for forward, training_step, etc.
+- Gradient clipping
+
+
+.. image:: ../_static/images/accelerator/overview.svg
+
+
```
Comment (edenlightning) [inline docs/source/extensions/plugins.rst:39]: Add a title
```suggestion
#################
Create a custom plugin
#################
```

**Review by edenlightning — COMMENTED** [review]
Nice work!

**Review by justusschock — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
Awesome!

**Review by s-rog — APPROVED** [review]
Was wondering about the empty plugin page on stable docs, this looks good!
For the lazy: https://108150-178626720-gh.circle-artifacts.com/0/html/extensions/plugins.html

**Comment by s-rog** [comment]
Perhaps we should also add a block for setting custom kwargs with the built in ddp plugins?

Similar to: https://pytorch-lightning.readthedocs.io/en/latest/benchmarking/performance.html#when-using-ddp-set-find-unused-parameters-false

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by edenlightning** [comment]
Looking good! The only thing i might suggest is using a more code like font for the class names but that's not super important
